modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java

```diff
@@ -2644,78 +2651,128 @@ private synchronized void stop0(boolean cancel, boolean waitForBackups) {
 
-                        AffinityTopologyVersion readyTopVer = cctx.topology().readyTopologyVersion();
+                    HashSet<UUID> nodesToExclude;
 
-                        if (readyTopVer.topologyVersion() < curTopVer) {
-                            if (log != null) {
-                                LT.warn(log, "Delaying node shutdown due to ongoing topology change [curTopVer=" +
-                                    curTopVer + ", readyTopVer=" + readyTopVer + "]");
-                            }
+                    try {
+                        originalNodesToExclude = metaStorage.read(GRACEFUL_SHUTDOWN_METASTORE_KEY);
+
+                        nodesToExclude = originalNodesToExclude != null ? new HashSet<>(originalNodesToExclude) :
+                            new HashSet<>();
+                    }
+                    catch (IgniteCheckedException e) {
+                        U.error(log, "Unable to read " + GRACEFUL_SHUTDOWN_METASTORE_KEY +
+                            " value from metastore.", e);
 
-                            blocked = true;
+                        continue;
+                    }
 
-                            break;
-                        }
+                    for (CacheGroupContext grpCtx : grid.context().cache().cacheGroups()) {
+                        if (grpCtx.systemCache())
+                            continue;
 
-                        if (cctx.config().getCacheMode() == LOCAL)
+                        if (grpCtx.isLocal())
                             continue;
 
-                        int locNodeMinNumOfParts = cctx.group().metrics().getLocalNodeMinimumNumberOfPartitionCopies();
-                        if (locNodeMinNumOfParts == 1) {
-                            if (cctx.config().getCacheMode() == PARTITIONED && cctx.config().getBackups() == 0) {
-                                if (log != null) {
-                                    LT.warn(log, "Ignoring potential data loss on cache without backups [name=" +
-                                        cache.getName() + "]");
+                        if (grpCtx.config().getCacheMode() == PARTITIONED && grpCtx.config().getBackups() == 0) {
+                            U.warn(log, "Ignoring potential data loss on cache without backups [name="
+                                + grpCtx.cacheOrGroupName() + "]");
+
+                            continue;
+                        }
+
+                        GridDhtPartitionFullMap fullMap = grpCtx.topology().partitionMap(false);
+
+                        int cacheSpecificAmountOfOwners = -1;
+
+                        if (fullMap != null) {
+                            nodesToExclude.retainAll(fullMap.keySet());
+
+                            int parts = grpCtx.topology().partitions();
+
+                            for (int part = 0; part < parts; part++) {
+                                Set<UUID> partIdealAssignment = grpCtx.affinity().idealAssignmentRaw().get(part).
```

Comment:
Topology / affinity cache can be updated concurrently. Possible solution: we can use topology read lock and / or check topology version of the affinity cache.

Reply:
There's no sense in such check cause we compare `grid.cluster().topologyVersion()` fixed at the beginning of the method and `grid.context().discovery().topologyVersion()`: `safeToStop = safeToStop && topVer == grid.context().discovery().topologyVersion();`

Reply:
Reworked, using compassion readyTopologyVersion of each cache group.